File: uva online judge/10036 - Divisibility.py
def read_test_case():
    n, k = (int(i) for i in input().split())
    return [abs(int(i)) % k for i in input().split()], k

def divisibility_dp(nums, k):
    n = len(nums)
    current = [False] * k
    current[0] = True
    for i in nums:
        previous, current = current, [False] * k
        for j in range(0, k):
            if previous[j]:
                current[abs(j + i) % k] = True
                current[abs(j - i) % k] = True
    return current[0]

if __name__ == '__main__':
    for i in range(int(input())):
        print('Divisible' if divisibility_dp(*read_test_case()) else 'Not divisible')

File: uva online judge/10036 - Divisibility.c
#include<stdio.h>
#include<string.h>
#define ABS(x) ((x)<0 ? -(x) : (x))
#define MAX_N 10000
#define MAX_K 100

int nums[MAX_N];

void read_test_case(int *n, int *k, int nums[]){
    int i;
    scanf("%d %d", n, k);
    for(i = 0; i < *n; i++){
        scanf("%d", &nums[i]);
        nums[i] = nums[i] % *k;
    }
}

int divisibility_dp(int n, int k, int nums[]){
    int i, j;
    char array1[MAX_K] = {1}, array2[MAX_K];
    char *current =  array1, *previous = array2, *tmp;
    for(i = 0; i < n; i++){
        tmp = current; current = previous; previous = tmp;
        memset(current, 0, k);
        for(j = 0; j < k; j++){
            if(previous[j]){
                current[ABS(j + nums[i]) % k] = 1;
                current[ABS(j - nums[i]) % k] = 1;
            }
        }
    }
    return current[0];
}

int main(){
    int t, n, k;
    scanf("%d", &t);
    while(t-- > 0){
        read_test_case(&n, &k, nums);
        printf(divisibility_dp(n, k, nums) ? "Divisible\n": "Not divisible\n") ;
    }
    return 0;
}